### DISCUSSION TITLE
What is the best way to add children to a parent node

### DISCUSSION THREAD

**Discussion by lubiah** [unifiedjs/unified#224] [Q&A] [answered]
Hi, 
I'm currently having a problem and I don't know the best way to go about it.  
Here's what I want to achieve .
* Use `unist-util-visit` to transverse a tree and find each code block.
* For each code block:
    * Loop over a theme array and generate a theme for the code block. (This means a node which has 3 themes will get three versions of itself) 

The problem now is
*  how to insert the nodes into the tree and how to tell visit to `SKIP` those nodes which I will add.
* I know I can use `parent.splice(index,1, nodes)` but how will I be able to generate the `Index` for those nodes which I'm adding

Here's a simplified version of the code

```javascript
visit(tree, "code", (node, index,parent )=>{
 const themes = ["a","b","c","d'];
 const newNodes = [];
for (const theme of themes) {
 //Do whatever you will do 
 newNodes.push(newCodeWithTheme)
}
})
```

So after it's done, how do I generate the index for these nodes and tell `visit` not to parse them

**Comment by ChristianMurphy** [comment]
Welcome @lubiah! 👋 
I'm not entirely clear if you are doing things in a single pass with the visitor, or multiple.
If you are doing a single pass, `unist-util-visit` include a `SKIP` action/return which can be used to skip the children of the current node, see https://github.com/syntax-tree/unist-util-visit#skip
If you are doing multiple passes, first you would need to detect if a node is/has a theme, then call the skip action to avoid re-processing it.

**Reply by lubiah** [reply]
What do you mean by single pass??

**Reply by ChristianMurphy** [reply]
Phrasing another way.
Do you have multiple `visitor`s in the same plugin or multiple plugins each with it's own `visitor`? or not?

**Reply by lubiah** [reply]
There's only one `visitor`

**Reply by lubiah** [reply]
There's only one `visitor`

**Reply by lubiah** [reply]
Let me make it this way, for each `Code` node that it finds, it duplicates it by the number of themes available. So let's say there are three themes. The current block in the visitor will get three version of itself.

**Reply by ChristianMurphy** [reply]
Okay, I'm not sure what you are asking then.
Could you rephrase? Maybe including a runnable example of what you are trying in a codesandbox/stackblitz?
Using https://github.com/unifiedjs/.github/blob/main/support.md as a reference for what to consider and include.

**Reply by lubiah** [reply]
Let me rephrase. 
When you read the docs for Shiki, the syntax highlighter, it says that in order to have two themes for both dark and light mode. You can render the same code block twice, each with a different theme. That's exactly what I'm doing.
Let me include a better example

```javascript
import { visit } from 'unist-util-visit';
const plugin = () => tree => {
  const themes = ['dracula', 'monokai', 'github-dark'];
  const blocksToHighlight = [];
  visit(tree, 'code', (node, index, parent) => {
    blocksToHighlight.push({node,index,parent});
  });

  for (const { node, index, parent } of blocksToHighlight){
    const highlightedNodes: Html[] = [];
    for (const theme of themes){
      highlightedNodes.push({
        type: "html",
        value: highlightedCode(theme)
      })
    }

    parent.children.splice(index,1,...highlightedNodes)
  }
```

This is the code and it works alright.
So each node is highlighted multiple times (by the number of themes), so the current node will have a "dracula","monokai" and "github-dark" version. 

Everything works as expected. The nodes are being inserted correctly but my worry is since this is done asynchronously (in my original code). Won't the visitor transverse the nodes I just added. How do I tell it to skip those ones.

Please  hope you understand me this time?

**Reply by ChristianMurphy** [reply]
> my worry is since this is done asynchronously 

Neither example appears to be asynchronous (using a callback which could have delayed resolution or using a `Promise`).

> Won't the visitor transverse the nodes I just added

In the original example yes.

> How do I tell it to skip those ones.

To avoid traversing the new children, inside the `visit` function, after you finish the `for` loop, return `SKIP` as I noted in my first reply https://github.com/syntax-tree/unist-util-visit#skip
That will skip the children of the node you just modified, but continue scanning for other code blocks elsewhere.

for example:

```ts
import { Root, Html } from 'mdast';
import { visit, SKIP } from 'unist-util-visit';

export const remarkShikiThemes = () => (tree: Root) => {
  const themes = ["dracula", "monokai", "github-dark"];
  visit(tree, "code", (node, index, parent) => {
    const highlightedNodes: Html[] = [];
    for (const theme of themes) {
      highlightedNodes.push({
        type: "html",
        value: highlightedCode(theme),
      });
    }

    parent.children.splice(index, 1, ...highlightedNodes);
    
    // skip children
    return SKIP;
  });
};
```

Also note the nodes being injected are `Html`, the guard on `visit` will skip these already.
Also also note, if you are injecting HTML, consider using a `rehype` plugin instead, like https://github.com/leafac/rehype-shiki

**Reply by lubiah** [reply]
Sorry, I failed to include this, in my original code, the highlighting is done asynchronously so I can't do that in the visit block. 

I use the visit blocks to store the nodes which I would like to highlight, then I'll iterate over those blocks and highlight them and since the addition of those nodes is done outside the visit function. I can't tell it to skip.
But since you said the visit function skips html, then I don't think it will be a problem again


But I did face another problem. Since the blocks are stored in an array, when I'm looping over them, only the first block works. This is because let's say for the index 0 in the stored array, I generate three code blocks for it, so when I use `parent.children.splice` to insert the nodes, for the index 1 in the stored array, it won't be able to insert the nodes at the correct position since I've added new nodes so the position for index 1 won't be same as it was. Is there any way to go about this?

Another question, why do you suggest I use rehype when remark does the job. Are there any disadvantages

**Reply by wooorm** [reply]
https://github.com/remarkjs/remark-rehype#what-is-this

**Reply by lubiah** [reply]
@wooorm , sorry if my question seems unjustifiable but I'm trying to create a `remark` plugin. How then can I use remark-rehype. I know I can use it when setting up unified to convert MD to html but when I export it the remark plugin, I'm not incharge of setting up the processor

**Reply by wooorm** [reply, marked as answer]
I linked to documentation that explains why both remark and rehype exist, and why you shouldn’t create a remark plugin when caring about HTML, but a rehype plugin, which is specifically made for dealing with HTML!

**Reply by lubiah** [reply]
@wooorm  you are right that I must use rehype when dealing with HTML. The reason why I chose to use remark instead was that with remark, I accomplished the same thing with less code. I saw an example of the same thing with rehype, the code was much plenty .I really don't do anything with HTML. I just insert it into the tree.
So am I making a mistake using remark?

**Reply by lubiah** [reply]
Anyways, I've decided to use rehype but my original question still remains unanswered

**Reply by wooorm** [reply]
I don’t really know what the original question is? Would be useful to have an example of your actual code, the actual problem you are facing, and to hear what you want to improve?

**Reply by lubiah** [reply]
Strangely, the problem has disappeared after re-writing everything to use `rehype` instead. I guess using `rehype` was the best choice after all. I'll go ahead and close the discussion.
Thanks for your time.
